sw/host/opentitantool/src/backend/mod.rs

```diff
@@ -29,21 +31,67 @@ pub struct BackendOpts {
 
     #[structopt(flatten)]
     verilator_opts: verilator::VerilatorOpts,
+
+    #[structopt(long, help = "Configuration files")]
+    conf: Option<String>,
 }
 
 #[derive(Error, Debug)]
 pub enum Error {
     #[error("unknown interface {0}")]
     UnknownInterface(String),
+    #[error("Missing interface (specify --conf or --interface)")]
+    MissingInterface(),
+    #[error("Missing interface, first included configuration file must specify")]
+    MissingInterfaceConf(),
+    #[error("Interface specified both on command line and in configuration (or twice in configuration files)")]
+    DoubleInterface(),
 }
 
 /// Creates the requested backend interface according to [`BackendOpts`].
-pub fn create(args: &BackendOpts) -> Result<Box<dyn Transport>> {
-    match args.interface.as_str() {
-        "" => Ok(Box::new(EmptyTransport)),
-        "verilator" => verilator::create(&args.verilator_opts),
-        "ultradebug" => ultradebug::create(args),
-        "cw310" => cw310::create(args),
-        _ => Err(Error::UnknownInterface(args.interface.clone()).into()),
+pub fn create(args: &BackendOpts) -> Result<TransportWrapper> {
+    let mut env = Option::None;
+    process_interface_spec(&mut env, &args.interface, &args)?;
+    for conf_file in &args.conf {
+        process_config_file(&mut env, &conf_file, args)?
+    }
+    match env {
+        Some(e) => Ok(e),
+        None => Err(Error::MissingInterface().into()),
+    }
+}
+
+fn process_interface_spec(env: &mut Option<TransportWrapper>, interface: &Option<String>, clunky_args: &BackendOpts) -> Result<()> {
+    let interface = match interface {
+        Some(i) => i,
+        None => { return Ok(()) }
+    };
+    let transport: Box<dyn Transport> = match interface.as_str() {
+        "" => Ok(Box::new(EmptyTransport) as Box<dyn Transport>),
+        "verilator" => verilator::create(&clunky_args.verilator_opts),
+        "ultradebug" => ultradebug::create(clunky_args),
+        "cw310" => cw310::create(clunky_args),
+        _ => Err(Error::UnknownInterface(interface.clone()).into()),
+    }?;
+    if env.is_some() {
+        return Err(Error::DoubleInterface().into());
+    }
+    env.insert(TransportWrapper::new(transport));
+    Ok(())
+}
```

Comment:
This function feels like it wants to be `fn process_interface_spec(interface: &str, args: &BackendOpts) -> Result<TransportWrapper>`.  Not having to unwrap `interface` would make this function flow more smoothly:

```
fn process_interface_spec(interface: &str, args: &BackendOpts) -> Result<TransportWrapper> {
  let transport = match interface { ... };
  Ok(TransportWrapper::new(transport))
```
Also, is there something particularly obnoxious about `BackendOpts` (you've named it `clunky_args` :) ).  Maybe we can fix it.  Personally, I'm not a big fan of the fact that it doesn't seem like I can use `flatten` together with `enable_if` to control which args are valid with which `interface`.  I have half a mind to just manually flatten and add the relevant `enable_if`s.

Reply:
I agree this piece of code is not as clean as I would have wanted.  The root cause is that with the current declaration, the TransportWrapper struct can only be constructed once we have a particular Transport implementation.  It may take parsing several levels of include directives in the configuration files, before we encounter the one .json file that has the interface field, allowing us to create a Transport implementation.  That is why it takes TransportWrapper as a mutable option reference.

I could change the interface as you propose, but then the calling code would need to recognize when the interface field is absent, as it is traversing the included configuration files, and only invoke the method when it is present, detecting if it is present in two or more places (file or command line).

I thought about making the transport field of TransportWrapper an Option, and create a TransportWrapper without one before parsing any configuration files, but I decided against, since that would make the code that operates on a fully constructed TransportWrapper more complicated to look at.

The best solution could be to introduce a TransportWrapperBuilder, which would have an optional Transport field.  I believe that there could be other advantages to doing resolution of multi-level aliases and other things at a time after all configuration files are parsed, which would fit will in a .build() method on such a Builder object.

I would like to gain a little more experience with a few more Transport implementations, and use of configuration files, before deciding, so maybe we can start with the code as is.